src/common/buffer.cc

```diff
@@ -917,43 +917,62 @@ static ceph::spinlock debug_lock;
     : iterator_impl(l, o, ip, po)
   {}
 
-  // copy data in
-  void buffer::list::iterator::copy_in(unsigned len, const char *src, bool crc_reset)
+  void buffer::list::iterator::copy_in(unsigned len, const ptr& bp)
   {
-    // copy
-    if (p == ls->end())
+    const auto orig_len = len;
+    if (len == 0) {
+      return;
+    }
+    list_iter_t prev_p; // = ls->before_begin();
+    if (p == ls->end()) {
       seek(off);
+    }
+    // punch hole
     while (len > 0) {
-      if (p == ls->end())
+      if (p == ls->end()) {
 	throw end_of_buffer();
-      
-      unsigned howmuch = p->length() - p_off;
-      if (len < howmuch)
-	howmuch = len;
-      p->copy_in(p_off, howmuch, src, crc_reset);
-	
-      src += howmuch;
-      len -= howmuch;
-      *this += howmuch;
+      }
+      const unsigned available_in_p = p->length() - p_off;
+      const unsigned take_from_p = std::min(available_in_p, len);
+      const unsigned residue_in_p = available_in_p - take_from_p;
```

Comment:
residue_in_p will be computed using the '-' operator on two unsigned. I'd rather make it a large signed variable (int64_t?), to make sure the 'if' below is robust.

Reply:
BTW - might it be a good idea to use ssize_t for all those lengths?